Modules/Loadable/Markups/SubjectHierarchyPlugins/qSlicerSubjectHierarchyMarkupsPlugin.cxx

```diff
@@ -764,6 +773,100 @@ void qSlicerSubjectHierarchyMarkupsPlugin::toggleSelectPoint()
   markupsNode->SetNthControlPointSelected(componentIndex, !markupsNode->GetNthControlPointSelected(componentIndex));
 }
 
+//-----------------------------------------------------------------------------
+void qSlicerSubjectHierarchyMarkupsPlugin::setTerminology()
+{
+  Q_D(qSlicerSubjectHierarchyMarkupsPlugin);
+
+  if (d->ViewMenuEventData.find("NodeID") == d->ViewMenuEventData.end())
+    {
+    qCritical() << Q_FUNC_INFO << ": No node ID found in the view menu event data";
+    return;
+    }
+  vtkMRMLScene* scene = qSlicerSubjectHierarchyPluginHandler::instance()->mrmlScene();
+  if (!scene)
+    {
+    qCritical() << Q_FUNC_INFO << ": Failed to access MRML scene";
+    return;
+    }
+  vtkMRMLSubjectHierarchyNode* shNode = scene->GetSubjectHierarchyNode();
+  if (!shNode)
+    {
+    qCritical() << Q_FUNC_INFO << ": Failed to access subject hierarchy node";
+    return;
+    }
+  vtkSlicerTerminologiesModuleLogic* terminologiesLogic = vtkSlicerTerminologiesModuleLogic::SafeDownCast(
+    qSlicerCoreApplication::application()->moduleLogic("Terminologies"));
+  if (!terminologiesLogic)
+    {
+    qCritical() << Q_FUNC_INFO << ": Failed to get Terminologies module logic";
+    return;
+    }
+
+  // Get markups node
+  QString nodeID = d->ViewMenuEventData["NodeID"].toString();
+  vtkMRMLNode* node = scene->GetNodeByID(nodeID.toUtf8().constData());
+  vtkMRMLMarkupsNode* markupsNode = vtkMRMLMarkupsNode::SafeDownCast(node);
+  if (!markupsNode)
+    {
+    qCritical() << Q_FUNC_INFO << ": Failed to get markups node by ID " << nodeID;
+    return;
+    }
+  markupsNode->CreateDefaultDisplayNodes();
+  vtkMRMLDisplayNode* displayNode = markupsNode->GetDisplayNode();
+  if (!displayNode)
+    {
+    qCritical() << Q_FUNC_INFO << ": Invalid display node";
+    return;
+    }
+
+  // Convert string list to VTK terminology entry. Do not check success, as an empty terminology is also a valid starting point
+  vtkNew<vtkSlicerTerminologyEntry> terminologyEntry;
+  const char* oldTerminologyAttribute = markupsNode->GetAttribute(vtkSegment::GetTerminologyEntryTagName());
+  std::string oldTerminologyString(oldTerminologyAttribute ? oldTerminologyAttribute : "");
+  terminologiesLogic->DeserializeTerminologyEntry(oldTerminologyString, terminologyEntry);
+
+  // If auto generated flags are not initialized, then set them to the default
+  // (color: on, name: off - this way color will be set from the selector but name will not)
+  bool nameAutoGenerated = false;
```

Comment:
When a terminology is selected then it should be possible to use it as node name. However, currently, when the popup is displayed and a term is selected the "Reset to default" button is grayed out for the name, so it is hard to make the selected terminology to be the node name (you need to type something, then the button becomes enabled, and you can click it ).

The reset to default button should be enabled (so that it can be clicked).

An alternative solution (nicer, but more work) would be to keep track if the node's name was automatically generated and if it was then make the terminology window automatically update the node name when a term is selected. Unfortunately, there is no mechanism to determine if a node name was automatically generated or not, so that would need to be implemented.

Reply:
Actually I think it is already implemented like what you say would be the nicer way. See `vtkSlicerTerminologiesModuleLogic::GetNameAutoGeneratedAttributeName()` in the code. The way I disabled it here, is because in the rest of this plugin it is also disabled, see `qSlicerSubjectHierarchyMarkupsPlugin::getDisplayColor` function. We can enable name auto generated, it was just a decision we made in the past (and I don't remember the reasons anymore and they are not commented so just decided to follow the convention).

Reply:
What we miss for the "nicer" solution is that we don't know if the node's name was automatically generated. In some cases it is very clear if the name is auto-generated, for example when you place a markup by clicking on the toolbar. It is also clear when a user manually renames a node that it is not auto-generated. However, when a module generates a name then it may or may not be very specific and sometimes you would want to keep that name, but sometimes it is so generic that you would prefer to replace it.

The bug in the current implementation is the initial state of the "reset name to default" button. It should become enabled when a term is selected and the term is different from the current node name.

Reply:
I see. Yes the name in this case is originally auto-generated.

I'll look into the reset name button state thing.

Reply:
The logic behind the reset button is quite complex actually, and I'm pretty sure if I change something to fix this scenario I'll brak at least another one. @lassoan so I tinhk we should discuss this.

(Btw a workaround for not being able to change the name is to change the name text and then the reset button becomes enabled)